src/compiler/checker.ts

```diff
@@ -24440,9 +24498,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return undefined;
     }
 
-    function isTupleTypeStructureMatching(t1: TupleTypeReference, t2: TupleTypeReference) {
+    function isTupleTypeStructureMatching(t1: TupleTypeReference, t2: TupleTypeReference, tupleStructureComparisonKind: TupleStructureComparisonKind) {
         return getTypeReferenceArity(t1) === getTypeReferenceArity(t2) &&
-            every(t1.target.elementFlags, (f, i) => (f & ElementFlags.Variable) === (t2.target.elementFlags[i] & ElementFlags.Variable));
+            every(t1.target.elementFlags, (f1, i) => {
+                const f2 = t2.target.elementFlags[i];
+                return f1 === f2 ||
+                    !!(tupleStructureComparisonKind & TupleStructureComparisonKind.MatchFixed && f1 & ElementFlags.Fixed && f2 & ElementFlags.Fixed) ||
+                    !!(tupleStructureComparisonKind & TupleStructureComparisonKind.MatchVariable && f1 & ElementFlags.Variable && f2 & ElementFlags.Variable);

```

Comment:
Why aren't these something like `f1 & ElementFlags.Fixed === f2 & ElementFlags.Fixed`, as the original code did?